### PR TITLE
Add support for Keras masking and causal masking

### PR DESCRIPTION
**Pull request by ageron**
This is a first go at adding Keras masking support and causal support to the `MultiHeadAttention` layer.

See the discussion in [#16248](https://github.com/keras-team/keras/issues/16248).

### CODE REVIEW

**Review by fchollet — COMMENTED** [review]
Thank you for the PR! For consistency, I would favor the same API as the one used in KerasNLP here: https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/transformer_decoder.py#L191

As in, a `use_causal_mask` argument in `call()`. It would not change the PR much otherwise.

**Comment by fchollet** [comment]
FYI @mattdangerw @chenmoneygithub

**Comment by ageron** [comment]
Hi @fchollet , thanks for reviewing this PR.

Note that the `keras.layers.Attention` layer uses an init parameter called `causal`, as noted by @haifeng-jin in #16248, so it seems to me that consistency within Keras is more important than consistency with Keras-NLP?

That said, I just made the change you requested, and I'm happy to push it if you prefer, I don't have a strong opinion about this: perhaps it's the `Attention` layer that should be updated to use `use_causal_mask` in the `call()` method.

**Comment by fchollet** [comment]
This is a good question -- we should make a choice one way or the other and standardize everything (Keras, KerasNLP) on that choice. Let me chat with the team and we'll figure out what to do.

**Comment by chenmoneygithub** [comment]
@fchollet To me `causal` is not the best argument name as it is too abstract. I did a codesearch and there are only a few usages of the `causal` argument of `tf.keras.layers.Attention` in google3, so arg name changing won't be too hard.

**Comment by ageron** [comment]
Sounds good. I just replaced `causal` in the constructor with `use_causal_mask` in the `call()` method.

**Comment by fchollet** [comment]
I also think it makes sense to have the argument in `call()` because that's where we pass the other masking-related arguments. So let's keep `use_causal_mask` in `call()`.

@chenmoneygithub could you please file an issue to change the `Attention` layer API to adopt this convention?

File: keras/layers/attention/multi_head_attention.py [hunk 1]
```diff
@@ -215,6 +215,9 @@ class MultiHeadAttention(Layer):
         training mode (adding dropout) or in inference mode (no dropout).
         Defaults to either using the training mode of the parent layer/model,
         or False (inference) if there is no parent layer.
+      use_causal_mask: A boolean to indicate whether to apply a causal mask to
+      prevent tokens from attending to future tokens (e.g., used in a decoder
```
Comment (fchollet) [inline keras/layers/attention/multi_head_attention.py:219]: Please add line indent here and below.

File: keras/layers/attention/multi_head_attention.py [hunk 2]
```diff
@@ -592,3 +606,87 @@ def call(
...
+        Args:
+          query: Projected query `Tensor` of shape `(B, T, N, key_dim)`.
+          key: Projected key `Tensor` of shape `(B, T, N, key_dim)`.
+          value: Projected value `Tensor` of shape `(B, T, N, value_dim)`.
+          attention_mask: a boolean mask of shape `(B, T, S)`, that prevents
+            attention to certain positions.
+          use_causal_mask: A boolean to indicate whether to apply a causal mask
+            to prevent tokens from attending to future tokens (e.g., used in a
+            decoder Transformer).
+        Returns:
```
Comment (fchollet) [inline keras/layers/attention/multi_head_attention.py:636]: Add line break above.

File: keras/layers/attention/multi_head_attention.py (hunk 1 above)
Comment (ageron) [reply keras/layers/attention/multi_head_attention.py]: Done

File: keras/layers/attention/multi_head_attention.py (hunk 2 above)
Comment (ageron) [reply keras/layers/attention/multi_head_attention.py:637]: Done

**Comment by chenmoneygithub** [comment]
@fchollet Sure!

**Review by fchollet — APPROVED** [review]
LGTM

**Comment by mattdangerw** [comment]
@ageron did some more testing a came across a few things.

1) I think we need to be more defensive about assuming mask types. We have real world usage where the mask (explicit or implicit) is floating point, which leads to errors when doing the bitwise `&`. Maybe we can just make sure we cast all masks to bool type before working with them? We should probably add a test for this too.

2) In an encoder where QKV are all the same, with an implicit mask, this code looks like it will make an attention mask with zeros along the bottom and along the right. I am used to seeing an attention mask with zeros only along one of those dimensions. This [colab](https://colab.sandbox.google.com/gist/mattdangerw/d25d83082112a5644f6e3688b32aa0d1/mask-differences.ipynb) the difference. I am reading your code correct? Do you know what the right approach here is?

**Comment by ageron** [comment]
Hi @mattdangerw ,
Sorry for the late response, got Covid in the family.
Thanks for the careful review, you make two good points.
I'll take care of the mask types.
However, regarding the padding difference, after some thought I'm not sure the difference matters, since it only affects masked out tokens, so normally these outputs will be ignored anyway.

For example, let's take a three word sentence padded to 5 tokens like this:

"I am happy <pad> <pad>"

With causal self-attention, we want "I" to attend only to "I", "am" should attend only to "I am", and "happy" should attend only to "I am happy". So far so good. Next, it doesn't really matter what both padding tokens attend to, since their output representation will be ignored anyway (downstream).

I tried to think of ways in which it could matter, but I couldn't find any, other than possibly a performance difference. Wdyt?

**Comment by ageron** [comment]
Hi @mattdangerw ,
Thanks for your feedback, I'll add the defensive casting + test by ~Wednesday, sounds good?

**Comment by ageron** [comment]
Hi @mattdangerw, I just added defensive casting to bool.

**Review by mattdangerw — APPROVED** [review]

**Comment by mattdangerw** [comment]
Still looking at this. We have one production failure from an oddly shaped Keras implicit mask we are trying to figure out.

Hopefully no more action needed here, but will ping if anything comes up!
